### DISCUSSION TITLE
General advice on multi window application

### DISCUSSION THREAD

**Discussion by eneoli** [Relm4/Relm4#590] [Q&A]
Hi there!

I'm currently working on a screenshot application similar to Flameshot and want to use relm4 as the GUI library.

The basic idea is that the application opens in full screen on any monitor, meaning a window is created for every monitor. Then the user can select the content they want and save/copy it to the clipboard. However, I'm facing a couple of issues:

1. If I understand correctly, the root of a relm4 application is a widget itself that has widgets as children (of the Widget type). My idea is to save a vector of windows as child widgets and create them on initialization. However, the root widget itself is a window. Is there a way to use some sort of "container" / dummy as the root just to manage the instantiation of the child windows? Or is there an even better option in relm4 to create multi-window applications?

2. A problem with my approach in (1) is that I won't be able to select across windows. My idea was to create a GTK DrawingArea and track mouse movements to handle the selection box properly. Are there any components that I could use that already provide this basic selection functionality?

My knowledge of GTK and relm4 is limited so I would greatly appreciate any suggestions or guidance on the best approach.

Thanks!

**Comment by AaronErhardt** [comment]
You could do the following:

1. Create a main component with an empty window and immediately hide it (if you use Relm4 from git, you can already use the [`visible_on_activate`](https://github.com/Relm4/Relm4/blob/main/relm4/src/app.rs#L82) method for hiding the window). 
2. Create child components from that component, similar to the example added in this PR (or maybe even use a factory): https://github.com/Relm4/Relm4/pull/562

Regarding your 2. point, unfortunately I don't quite understand what you want to archive exactly.

**Reply by eneoli** [reply]
Thank you for your response! It seems a bit peculiar to create a window and promptly hide it, but if that's how to do it, I'm happy with that ^^.

I apologize for any lack of clarity in my second point.

My objective is to achieve something akin to [this](https://raw.githubusercontent.com/flameshot-org/flameshot/master/data/img/preview/animatedUsage.gif).

I plan to create a window for each monitor displaying a screenshot of it. Subsequently, I aim to track mouse movements to obtain the size and position of the bounding box. The box itself, I intend to render using a GTK DrawingArea. Thus, my concern is whether, while holding down the mouse button, I can move to a different monitor and still track the events on the monitor I entered without releasing the mouse press.

I hope this provides a clearer understanding. Is this approach viable, or are there more effective methods utilizing GTK/Relm features?

**Reply by AaronErhardt** [reply]
>  It seems a bit peculiar to create a window and promptly hide it, but if that's how to do it, I'm happy with that ^^.

You can do it differently as well. For example the settings list example shows how to launch components from more or less raw gtk-rs: https://github.com/Relm4/Relm4/blob/main/examples/settings_list.rs#L21
That requires a bit more knowledge about GTK though.

> My objective is to achieve something akin to this.

I'm not very familiar with the GTK API regarding this and how they integrate with portals (for screenshots). The idea sound reasonable, but I could imagine that the seamless transition from one window to the other doesn't work really well. I also don't know whether it's easy to create a window that spans across monitors (which would be another solution, I guess). Honestly, the people in the [gtk-rs Matrix room](https://matrix.to/#/#rust:gnome.org) are likely to give you better hints.
If you have a question about state management and Relm4, feel free to comment here again or ask in [our Matrix room](https://matrix.to/#/#relm4:matrix.org).
